lib/autoupdate/rolloutcontroller/client_test.go

```diff
@@ -0,0 +1,188 @@
+	return m.updateAutoUpdateAgentRollout.handle(ctx, rollout)
+}
+
+func (m mockClient) DeleteAutoUpdateAgentRollout(ctx context.Context) error {
+	return m.deleteAutoUpdateAgentRollout.handle(ctx)
+}
+
+func (m mockClient) checkIfEmpty(t *testing.T) {
+	require.True(t, m.getAutoUpdateConfig.isEmpty(), "Get autoupdate_config mock not empty")
+	require.True(t, m.getAutoUpdateVersion.isEmpty(), "Get autoupdate_version mock not empty")
+	require.True(t, m.getAutoUpdateAgentRollout.isEmpty(), "Get autoupdate_agent_rollout mock not empty")
+	require.True(t, m.createAutoUpdateAgentRollout.isEmpty(), "Create autoupdate_agent_rollout mock not empty")
+	require.True(t, m.updateAutoUpdateAgentRollout.isEmpty(), "Update autoupdate_agent_rollout mock not empty")
+	require.True(t, m.deleteAutoUpdateAgentRollout.isEmpty(), "Delete autoupdate_agent_rollout mock not empty")
+}
+
+func newMockClient(t *testing.T, stubs mockClientStubs) *mockClient {
+	// Fail early if there's a mismatch
+	require.Equal(t, len(stubs.createRolloutAnswers), len(stubs.createRolloutExpects), "invalid stubs, create validations and answers slices are not the same length")
+	require.Equal(t, len(stubs.updateRolloutAnswers), len(stubs.updateRolloutExpects), "invalid stubs, update validations and answers slices are not the same length")
+
+	return &mockClient{
+		getAutoUpdateConfig:          &getHandler[*autoupdate.AutoUpdateConfig]{t, stubs.configAnswers},
+		getAutoUpdateVersion:         &getHandler[*autoupdate.AutoUpdateVersion]{t, stubs.versionAnswers},
+		getAutoUpdateAgentRollout:    &getHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.rolloutAnswers},
+		createAutoUpdateAgentRollout: &createUpdateHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.createRolloutExpects, stubs.createRolloutAnswers},
+		updateAutoUpdateAgentRollout: &createUpdateHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.updateRolloutExpects, stubs.updateRolloutAnswers},
+		deleteAutoUpdateAgentRollout: &deleteHandler{t, stubs.deleteRolloutAnswers},
+	}
+}
+
+type mockClientStubs struct {
+	configAnswers        []callAnswer[*autoupdate.AutoUpdateConfig]
+	versionAnswers       []callAnswer[*autoupdate.AutoUpdateVersion]
+	rolloutAnswers       []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	createRolloutAnswers []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	createRolloutExpects []require.ValueAssertionFunc
+	updateRolloutAnswers []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	updateRolloutExpects []require.ValueAssertionFunc
+	deleteRolloutAnswers []error
+}
+
+type callAnswer[T any] struct {
+	result T
+	err    error
+}
+
+// getHandler is used in a mock client to answer get resource requests during tests.
+// It takes a list of answers and errors and will return them when invoked.
+// If there are no stubs left it fails the test.
+type getHandler[T protoadapt.MessageV1] struct {
```

Comment:
Do we need the message v1 support? Isn't this supposed to be brand new types? 😢 

```suggestion
type getHandler[T proto.Message] struct {
```